File: lab_2_python/main_2.py
import heapq


n = int(input())
p = list(map(int, input().split()))
k = int(input())
line = [list() for i in range(n)]
for i in range(k):
    x, y = map(int, input().split())
    x -= 1
    y -= 1
    line[x].append(y)
    line[y].append(x)
heap = []
b = [False] * n
heapq.heappush(heap, 0)
ans = -1
while len(heap) != 0:
    z = heapq.heappop(heap)
    c = z // 100
    v = z % 100
    if b[v] == True:
        continue
    b[v] = True
    for u in line[v]:
        if not b[u]:
            heapq.heappush(heap, (c + p[v]) * 100 + u)
        if u == n - 1:
            if ans == -1:
                ans = c + p[v]
            else:
                ans = min(ans, c + p[v])
if n == 1:
    print(0)
else:
    print(ans)

File: lab_2_kotlin/src/main_2.kt
import java.util.TreeSet
import kotlin.math.min

// 7 Informatics
fun main(args: Array<String>) {
    var n = readLine()!!.toInt()
    var p = readLine()!!.trim().split(" ").map{it.toInt()}
    var k = readLine()!!.toInt()
    var line = Array(k, {ArrayList<Int>()})
    for (i in 0 until k) {
        var (x, y) = readLine()!!.split(" ").map{it.toInt()}
        x--
        y--
        line[x].add(y)
        line[y].add(x)
    }
    var heap = TreeSet<Pair<Int, Int>>({a, b -> a.first.compareTo(b.first)})
    var b = Array(n, {false})
    heap.add(Pair(0, 0))
    var ans = -1
    while (heap.size != 0) {
        var (c, v) = heap.pollFirst()
        if (b[v])
            continue
        b[v] = true
        for (u in line[v]) {
            if (!b[u])
                heap.add(Pair(c + p[v], v))
            if (u == n - 1)
                if (ans == -1)
                    ans = c + p[v]
                else
                    ans = min(ans, c + p[v])
        }
    }
    if (n == 1)
        print(0)
    else
        print(ans)
}